File: android/src/main/kotlin/com/zeticai/zetic_mlange_flutter/core/MultipleInstances.kt
package com.zeticai.zetic_mlange_flutter.core

import io.flutter.plugin.common.MethodChannel.Result

abstract class MultipleInstances<T> {
    private val instances: MutableMap<String, T> = mutableMapOf()

    protected fun createInstance(instanceId: String, instance: T) {
        instances[instanceId] = instance
    }

    protected fun removeInstance(instanceId: String): T? {
        return instances.remove(instanceId)
    }

    protected fun getInstance(instanceId: String, result: Result): T {
        val instance = instances[instanceId]
        if (instance == null) {
            result.error(
                "NO_INSTANCE",
                "Instance not created yet",
                null
            )
            throw IllegalArgumentException("Instance not created yet")
        }
        return instance
    }
}

File: ios/Classes/core/MultipleInstances.swift
import Flutter

open class MultipleInstances<T> {
    private var instances = [String: T]()

    open func createInstance(instanceId: String, instance: T) {
        instances[instanceId] = instance
    }

    open func removeInstance(instanceId: String) -> T? {
        return instances.removeValue(forKey: instanceId)
    }

    open func getInstance(instanceId: String, result: FlutterResult) -> T {
        guard let instance = instances[instanceId] else {
            result(
                FlutterError(
                    code: "NO_INSTANCE",
                    message: "Instance not created yet",
                    details: nil
                )
            )
            fatalError("Instance not created yet")
        }
        return instance
    }
}
